plugins/outputs/websocket/websocket.go

```diff
@@ -0,0 +1,216 @@
+package websocket
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/outputs"
+	"github.com/influxdata/telegraf/plugins/serializers"
+	"github.com/influxdata/telegraf/testutil"
+
+	ws "github.com/gorilla/websocket"
+)
+
+var sampleConfig = `
+  ## URL is the address to send metrics to. Make sure ws or wss scheme is used.
+  url = "ws://127.0.0.1:8080/telegraf"
+
+  ## Timeouts (make sure read_timeout is larger than server ping interval or set to zero).
+  # connect_timeout = "30s"
+  # write_timeout = "30s"
+  # read_timeout = "30s"
+
+  ## Optionally turn on using text data frames (binary by default).
+  # use_text_frames = false
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+
+  ## Data format to output.
+  ## Each data format has it's own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
+  # data_format = "influx"
+
+  ## Additional HTTP Upgrade headers
+  # [outputs.websocket.headers]
+  #   Authorization = "Bearer <TOKEN>"
+`
+
+const (
+	defaultConnectTimeout = 30 * time.Second
+	defaultWriteTimeout   = 30 * time.Second
+	defaultReadTimeout    = 30 * time.Second
+)
+
+// WebSocket can output to WebSocket endpoint.
+type WebSocket struct {
+	URL            string            `toml:"url"`
+	ConnectTimeout config.Duration   `toml:"connect_timeout"`
+	WriteTimeout   config.Duration   `toml:"write_timeout"`
+	ReadTimeout    config.Duration   `toml:"read_timeout"`
+	Headers        map[string]string `toml:"headers"`
+	UseTextFrames  bool              `toml:"use_text_frames"`
+	Logger         telegraf.Logger   `toml:"-"`
+	tls.ClientConfig
+
+	conn       *ws.Conn
+	serializer serializers.Serializer
+}
+
+// SetSerializer implements serializers.SerializerOutput.
+func (w *WebSocket) SetSerializer(serializer serializers.Serializer) {
+	w.serializer = serializer
+}
+
+// Description of plugin.
+func (w *WebSocket) Description() string {
+	return "Generic WebSocket output writer."
+}
+
+// SampleConfig returns plugin config sample.
+func (w *WebSocket) SampleConfig() string {
+	return sampleConfig
+}
+
+var errInvalidURL = errors.New("invalid websocket URL")
```

Comment:
You might want to use the text directly below instead of defining a new error.

Reply:
It's used in tests to not match error by text, not sure what's bad in this since it's not exported?

Reply:
There's nothing bad about it. It's more a matter of taste as I prefer to have the error text directly at the location the error triggers. This helps debugging, but you can keep the definition (hence I wrote 'might' :-)) if you prefer.

Reply:
Got it, let's leave as it's now then - should not have any impact.

Reply:
@srebhan the advantage of this is you can use `if errors.Is(err, errInvalidURL) {`, but that doesn't seem to be used here.